### ISSUE TITLE
max-messages-per-poll and max-fetch-size not working as expected

### ISSUE THREAD

**Issue by SimoneGiusso** [spring-projects/spring-integration#9124]
**In what version(s) of Spring Integration are you seeing this issue?**

6.2.4

**Describe the bug**

According to the [documentation](https://docs.spring.io/spring-integration/reference/sftp/max-fetch.html) the following combination of values `max-messages-per-poll=-1` and `max-fetch-size=1` should have the following effect: the sftp synchronizer fetch **all** the files from sftp server but **one** at time. As soon as a file is downloaded, a message with the file as payload, is fired in the channel.

However with this setting, in a single pool, the synchronizer only fetch 1 file. 

**To Reproduce**

Sftp server containing more than 1 files and `max-messages-per-poll=-1` and `max-fetch-size=1`

**Expected behavior**

Fetch all the files and sent a message as soon as a file is downloaded.

**Comment by artembilan** [comment]
Sorry, I need more explanation what is expected.
The `max-fetch-size=1` means only transfer to local directory one of per fetch:
```
filteredFiles = filteredFiles.stream()
					.limit(maxFetchSize)
					.collect(Collectors.toList());
```
Now the `max-messages-per-poll=-1` is like this:
```
					while (this.initialized && (this.maxMessagesPerPoll <= 0 || count < this.maxMessagesPerPoll)) {
						if (this.maxMessagesPerPoll == 0) {
							logger.info("Polling disabled while 'maxMessagesPerPoll == 0'");
							break;
						}
						if (pollForMessage() == null) {
							break;
						}
						count++;
					}
```
That `pollForMessage()` calls:
```
	public final AbstractIntegrationMessageBuilder<File> doReceive(int maxFetchSize) {
		AbstractIntegrationMessageBuilder<File> messageBuilder = this.fileSource.doReceive();
		if (messageBuilder == null) {
			this.synchronizer.synchronizeToLocalDirectory(this.localDirectory, maxFetchSize);
			messageBuilder = this.fileSource.doReceive();
		}
```
So, as long as you ask for only one file to `synchronizeToLocalDirectory` there is no way to cache more of them in the local dir.

Would you mind to elaborate more what would you like to see?

**Comment by SimoneGiusso** [comment]
I set the `max-messages-per-poll` (infinity) to a higher value than `max-fetch-size` (1). Now when `max-messages-per-poll` > `max-fetch-size` this should be the behaviour, according to the documentation:

> max-messages-per-poll=2 and max-fetch-size=1: The adapter fetches one file, emits it, fetches the next file, and emit it. Then it sleeps until the next poll.

According to the description, with my configuration, I would expect that the adapter fetches one file, emits it, fetches the next file, emit it, etc... until all the files have been fetched. However the behaviour I'm seeing is the following:

adapter fetches one file, emits it

That's it!!! No more files are downloaded to the local repository, hence no more messages are emitted.

**Comment by artembilan** [comment]
> No more files are downloaded to the local repository, hence no more messages are emitted.

Then something is off with the filter for that channel adapter.
The logic there is like this:
```
			boolean filteringOneByOne = haveFilter && this.filter.supportsSingleFileFiltering();
			List<F> filteredFiles = applyFilter(files, haveFilter, filteringOneByOne, maxFetchSize);
```

May we have your whole configuration for that channel adapter to investigate, please?

**Comment by SimoneGiusso** [comment]
It's not the filter. If I don't' set the max fetch size all the files are downloaded

**Comment by artembilan** [comment]
I cannot reproduce it in my testing environment:
```
	@Test
	public void testSftpInboundFlow() {
		QueueChannel out = new QueueChannel();
		IntegrationFlow flow = IntegrationFlow
				.from(Sftp.inboundAdapter(sessionFactory())
								.preserveTimestamp(true)
								.remoteDirectory("sftpSource")
								.regexFilter(".*\\.txt$")
								.localFilenameExpression("#this.toUpperCase() + '.a'")
								.localDirectory(getTargetLocalDirectory())
								.remoteComparator(Comparator.comparing(SftpClient.DirEntry::getFilename))
								.maxFetchSize(1),
						e -> e.id("sftpInboundAdapter").poller(Pollers.fixedDelay(100).maxMessagesPerPoll(-1)))
				.channel(out)
				.get();
		IntegrationFlowRegistration registration = this.flowContext.registration(flow).register();
		Message<?> message = out.receive(10_000);
		assertThat(message).isNotNull();
		Object payload = message.getPayload();
		assertThat(payload).isInstanceOf(File.class);
		File file = (File) payload;
		assertThat(file.getName()).isEqualTo(" SFTPSOURCE1.TXT.a");
		assertThat(file.getAbsolutePath()).contains("localTarget");

		message = out.receive(10_000);
		assertThat(message).isNotNull();
		file = (File) message.getPayload();
		assertThat(file.getName()).isIn("SFTPSOURCE2.TXT.a");
		assertThat(file.getAbsolutePath()).contains("localTarget");

		registration.destroy();
	}
```
Both files are downloaded as expected: one by one.

Apparently you indeed need to share with me some simple project to reproduce.

Thanks

**Comment by SimoneGiusso** [comment]
@artembilan here you are -> https://github.com/SimoneGiusso/spring-integration-sftp-demo/tree/main

The test run against a real sftp server. The test should not pass since only one file is downloaded instead of the two available which satisfy the filter.

I also noticed this warning:

```
2024-05-07T22:18:10.611+02:00  WARN 79505 --- [   scheduling-1] o.s.i.s.i.SftpInboundFileSynchronizer    : The remote file 'File2_data.csv[-rw-r--r--    1 501      dialout         0 May  7 20:17 File2_data.csv]: type=1;size=0;uid=501;gid=20;perms=0x81a4;flags=[Size, UidGid, Perms, AccessTime, ModifyTime];owner=null;group=null;aTime=2024-05-07T20:18:10Z;cTime=null;mTime=2024-05-07T20:17:11Z;extensions=[]' has not been transferred to the existing local file '/tmp/extractions/File2_data.csv'. Consider removing the local file.
2024-05-07T22:18:10.611+02:00  WARN 79505 --- [   scheduling-1] o.s.i.s.i.SftpInboundFileSynchronizer    : The remote file 'File1_data.csv[-rw-r--r--    1 501      dialout         0 May  7 20:17 File1_data.csv]: type=1;size=0;uid=501;gid=20;perms=0x81a4;flags=[Size, UidGid, Perms, AccessTime, ModifyTime];owner=null;group=null;aTime=2024-05-07T20:18:10Z;cTime=null;mTime=2024-05-07T20:17:11Z;extensions=[]' has not been transferred to the existing local file '/tmp/extractions/File1_data.csv'. Consider removing the local file.
```

I don't think is excepted. Something is wrong.

**Comment by artembilan** [comment]
I think the problem is really here:
```
fileSynchronizer.setFilter(SftpSimplePatternFileListFilter("*_data.csv"))
```
You have to compose it with once filter , like we do in the framework for DSL: 
https://github.com/spring-projects/spring-integration/blob/8b88668fbeb842ec9003ebb1363981bd04948404/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/dsl/SftpInboundChannelAdapterSpec.java#L74

**Comment by SimoneGiusso** [comment]
I don't understand. The filter is not applied to each file to decide which one to download? Why max-fetch-size influence the behavior of the filter?

**Comment by artembilan** [comment]
Yes, that’s the right question. I cannot run your application since I don’t have real SFTP server on my Windows. I need to develop something similar to be able to reproduce. I guess my embedded MINA should do the trick. The point is that the problem is somewhere in app or in the framework as you claim.

**Comment by SimoneGiusso** [comment]
I think it should work on windows as well. The ssh server is run inside a docker container. You shouldn't doing any anything. TestContainer should pull and run the image for you.

**Comment by artembilan** [comment]
So, yeah... I thought about Docker image, but apparently that's a bit involved with your application.
I made a change in my unit test against MINA like this:
```
.filter(new SftpRegexPatternFileListFilter(".*\\.txt$"))
```
Fully similar to what you have so far in your app with a:
```
fileSynchronizer.setFilter(SftpSimplePatternFileListFilter("*_data.csv"))
```
And I see exact behavior as you explained.
Now if I do this in my test:
```
								.filter(
										new CompositeFileListFilter<>(
												List.of(new SftpRegexPatternFileListFilter(".*\\.txt$"),
														new SftpPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(),
																"sftpMessageSource"))))
```
It starts working as expected.

Now what is going on.

If you don't provide `SftpPersistentAcceptOnceFileListFilter`, then downloaded files are not going to be filtered out, and since the list of remote files is always the same, only the first remote file is returned.
That is according to the `.maxFetchSize(1)`.
Now, since we are in the synchronization mode, the `LocalFileReadingMessageSource` is involved to pull files from that local dir.
And it comes with its own:
```
this.localFileListFilter =
					new FileSystemPersistentAcceptOnceFileListFilter(new SimpleMetadataStore(), getComponentName());
```
So, same first file is trying to be downloaded (according to missed `SftpPersistentAcceptOnceFileListFilter`), but it is already there is the local dir.
That's why we see that `has not been transferred to the existing local file` warning.
And since this local file has been filtered out before (according to the `localFileListFilter`), nothing is going to be emitted.

So, the fix for your application is to compose your `SftpSimplePatternFileListFilter` with that `SftpPersistentAcceptOnceFileListFilter`.

As a fix for this issue we probably can take a look into the mentioned documentation and add respective recommendation.

**Comment by SimoneGiusso** [comment]
What you have explained seems quite far from what I understood reading the documentation. This is how I interpreted these two parameters:

- `max-messages-per-poll`: indicates the maximum number of messages to be emitted from the local repo to the channel in each `polling-interval`, hence the maximum number of messages downloaded and processed at every `polling-interval`.
- `max-fetch-size`: indicates the number of messages that have to be fetched before emitting message (`Message<FIle>`) to the channel. 

This is instead how I imaged the workflow:

1. ssh client connect to the repository, list the files
2. the downloading starts for the files passing the `filter`s  set in `SftpInboundFileSynchronizer` until number of files downloaded = `max-fetch-size`.
3. the files, downloaded in the local repo are now emitted. Also here a filter is applied. This time the `LocalFileReadingMessageSource` filters. The files downloaded in this `polling-interval` which pass these latter filters are sent to the channel. If numbers of files in this `polling-interval` > `max-messages-per-poll` this `polling-interval` finishes here. Otherwise return to point 1, restarting from the files that passes the `SftpInboundFileSynchronizer`'s filter but that have not been downloaded in this specific `polling-interval` yet. 

Now, what I want to achieve is the following: I want to connect to the remote repo, download **all the files** that pass the `SftpInboundFileSynchronizer`'s filters, one by one, i.e. as soon as the file is downloaded to the local repo, is sent to the `sftpChannel`.

In my example I just use a filter on the filename. I don't care that at the next iteration (next `polling-interval`) the same files will be downloaded and processed. I just wanted to understand how a single iteration works. 

For this reason I chose `max-fetch-size` = 1 (download one file at time) and `max-messages-per-poll` = infinity (process all the files in a single `polling-interval`.

Given my example where I have 3 files in sftp, two of which have a valid file name I would expect this workflow:

1. 3 files are present in Sftp <Valid1_data.csv, Valid2_data.csv, Invalid.csv>
2. Filter applied : <Valid1_data.csv, Valid2_data.csv>
3. Download first file: `Valid1_data.csv`. File remained for download: `Valid2_data.csv`
4. `Valid1_data.csv` downloaded to local repo
5. `Valid1_data.csv` file emitted to `sftpChannel`
6. download second file: `Valid2_data.csv`. File remained for download: `zero`
7. `Valid2_data.csv ` downloaded to local repo
8. `Valid2_data.csv` file emitted to `sftpChannel`
9. Finish iteration: no more files available

At the next pooling, since there is only a simple filter on the name, the same exact thing will happen again from point 1 to point 9.

**Comment by artembilan** [comment]
OK. Let's see what `max-fetch-size` is exactly about from the code perspective and the logic behind it:
```
F[] files = session.list(remoteDirectory);
List<F> filteredFiles = applyFilter(files, haveFilter, filteringOneByOne, maxFetchSize);
```

So, the `list` returns all the file names in the remote directory, in some order.
Usually it is alphabetic.
Then we apply the filter for those names.
And since we ask only for a single file, the first one accepted is returned.
Since you don't track processed files, it is not a surprise that we always get the same first file, but since it has been processed by the `FileSystemPersistentAcceptOnceFileListFilter`, nothing is going to be emitted as a message.

For the use-case you describe apparently a custom `FileListFilter` could be implemented, but that's already outside of the framework.

You can have something similar to the `SftpPersistentAcceptOnceFileListFilter`, but reset its cache on each poll.
See `ReceiveMessageAdvice`.

However same has to be applied to the `localFilter` as well. This one also caches already filtered files to avoid duplication.

**Comment by SimoneGiusso** [comment]
Thanks a lot for your explanations. Now it's more clear. Apparently everything is working as expected but maybe this:

> And since we ask only for a single file, the first one accepted is returned.
> Since you don't track processed files, it is not a surprise that we always get the same first file

Should be mentioned in the documentation, when specifying a value for `max-fetch-size`. I started to get confused in fact, when I specified `max-fetch-size`. And this behaviour will happen every time the number of files passing the filter is > `max-fetch-size` because without `SftpPersistentAcceptOnceFileListFilter` only `max-fetch-size` files will be downloaded.
